File: 2-Add-Two-Numbers/add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        head = last = None
        has_carry = False;
        while(l1 or l2):
            v1 = l1.val if l1 else 0
            v2 = l2.val if l2 else 0
            value = v1+v2
            if has_carry:
                value += 1
            if value >= 10:
                has_carry = True
                value -= 10
            else:
                has_carry = False;
            if head is None:
                head = ListNode(value)
                last = head
            else:
                last.next = ListNode(value)
                last = last.next
            
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None
            
        if has_carry:
            last.next = ListNode(1)
            
        return head
        

File: 2-Add-Two-Numbers/add-two-numbers.cpp
#include <iostream>
using namespace std;

struct ListNode {
      int val;
      ListNode *next;
      ListNode(int x) : val(x), next(NULL) {}
  };

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode node(0);
        ListNode* tmp  = &node;
        ListNode* head = tmp; 
        bool carray = false;
        while(l1 || l2){
            
            //calculate value
            int value = (l1?l1->val:0) + (l2?l2->val:0);
            if(carray){
                value += 1;
            }
            if(value >= 10){
                value -= 10;
                carray = true;
            }else{
                carray = false;
            }      
           
            //update tmp
            if(l1){
                l1->val = value;
                tmp->next= l1;
                tmp = tmp->next;
            }else{
                if(l2){
                    l2->val = value;
                    tmp->next = l2;
                    tmp = tmp->next;
                }
            }
            
            //update l1 & l1
            if(l1){
                l1=l1->next;
            }
            if(l2){
                l2 = l2->next;
            }
        }
         if(carray){
            tmp->next = new ListNode(1);
        }
        return head->next;
    } 
};

int main(){




    return 0;
}
 
 